### PR TITLE
Manage scatter rather than try to eliminate it

### PR DESCRIPTION
**Pull request by beckerg**
Signed-off-by: Greg Becker <gbecker@micron.com>

### CODE REVIEW

File: lib/cn/csched_sp3.c [hunk 1]
```diff
@@ -913,7 +913,10 @@ sp3_scatter(const struct cn_tree_node *tn)
     list_for_each_entry(le, &tn->tn_kvset_list, le_link) {
         uint vgroups = kvset_get_vgroups(le->le_kvset);
 
-        if (vgroups > 1)
+        /* Include the youngest kvsets with no scatter, but exclude
```
Comment (tristan957) [inline lib/cn/csched_sp3.c:916]: Comment wrapping!

File: lib/cn/csched_sp3.c [hunk 2]
```diff
@@ -937,18 +940,22 @@ sp3_dirty_node_locked(struct sp3 *sp, struct cn_tree_node *tn)
...
 
     jobs = atomic_read_acq(&tn->tn_busycnt);
-    nkeys = cn_ns_keys(&tn->tn_ns);
 
     nkvsets_total = cn_ns_kvsets(&tn->tn_ns);
     nkvsets = nkvsets_total - (jobs & 0xffffu);
     jobs >>= 16;
 
     if (tn->tn_parent) {
+        uint64_t nkeys = cn_ns_keys(&tn->tn_ns);
```
Comment (tristan957) [inline lib/cn/csched_sp3.c:949]: const

File: lib/cn/csched_sp3.c [hunk 3]
```diff
@@ -937,18 +940,22 @@ sp3_dirty_node_locked(struct sp3 *sp, struct cn_tree_node *tn)
...
+
+        scatter = sp3_scatter(tn);
 
         /* RBT_LI_LEN: internal and leaf nodes sorted by number of kvsets.
+         * We use inverse scatter as a secondary discriminant so as to
+         * prefer scatter jobs over kcompactions when scatter is high.
          */
         if (nkvsets >= sp->thresh.llen_runlen_min && jobs < 1) {
-            uint64_t weight = (nkvsets << 32) | nkeys;
+            uint64_t weight = (nkvsets << 32) | (UINT32_MAX - scatter);
```
Comment (tristan957) [inline lib/cn/csched_sp3.c:958]: const

File: lib/cn/csched_sp3.c [hunk 4]
```diff
@@ -1036,18 +1029,27 @@ sp3_dirty_node_locked(struct sp3 *sp, struct cn_tree_node *tn)
                 sp3_node_remove(sp, spn, RBT_L_PCAP);
             }
         }
+    } else {
+
+        /* If this root node is ready to spill then ensure it's on the list
+         * in FIFO order, retaining its current position if it's already on
+         * the list.  List order is otherwise managed by sp3_check_roots().
+         */
+        if (nkvsets >= sp->thresh.rspill_kvsets_min && jobs < 3) {
```
Comment (tristan957) [inline lib/cn/csched_sp3.c:1038]: Where do the jobs limits of 1 and 3 come from?

File: lib/cn/csched_sp3_work.c [hunk 5]
```diff
@@ -419,6 +422,11 @@ sp3_work_leaf_len(
         *rule = CN_CR_LLONG;
 
         list_for_each_entry_reverse (le, head, le_link) {
+            uint vgroups = kvset_get_vgroups(le->le_kvset);
```
Comment (tristan957) [inline lib/cn/csched_sp3_work.c:425]: const :)

File: lib/cn/csched_sp3_work.c [hunk 6]
```diff
@@ -419,6 +422,11 @@ sp3_work_leaf_len(
...
+
+            if (scatter + vgroups > 1)
```
Comment (tristan957) [inline lib/cn/csched_sp3_work.c:427]: If this logic is duplicated, would it be better to have a static inline somewhere?

File: lib/kvdb/kvdb_rparams.c [hunk 7]
```diff
@@ -806,7 +806,7 @@ static const struct param_spec pspecs[] = {
         .ps_stringify = param_default_stringify,
         .ps_jsonify = param_default_jsonify,
         .ps_default_value = {
-            .as_uscalar = 32,
+            .as_uscalar = 64,
```
Comment (tristan957) [inline lib/kvdb/kvdb_rparams.c:809]: Test needs to change

File: lib/cn/csched_sp3.c (hunk 4 above)
Comment (beckerg) [reply lib/cn/csched_sp3.c:1038]: We can never run more than one job at a time on a leaf node.
We have at most 5 threads available for root node processing, but we choose to not let more than 3 of those work on any given root node.  I could probably make a #define for that..

File: lib/cn/csched_sp3_work.c (hunk 6 above)
Comment (beckerg) [reply lib/cn/csched_sp3_work.c:427]: Sure, I'll do that...

File: lib/kvdb/kvdb_rparams.c (hunk 7 above)
Comment (beckerg) [reply lib/kvdb/kvdb_rparams.c:809]: Thanks! I only ran the smoke tests so far..

File: lib/cn/csched_sp3.c (hunk 2 above)
Comment (beckerg) [reply lib/cn/csched_sp3.c:949]: done

File: lib/cn/csched_sp3.c (hunk 3 above)
Comment (beckerg) [reply lib/cn/csched_sp3.c:958]: done, made them all const where applicable

File: lib/cn/csched_sp3.c (hunk 4 above)
Comment (beckerg) [reply lib/cn/csched_sp3.c:1038]: Added a comment instead.

File: lib/cn/csched_sp3_work.c (hunk 5 above)
Comment (beckerg) [reply lib/cn/csched_sp3_work.c:425]: eliminated due to your following suggestion

File: lib/cn/csched_sp3_work.c (hunk 6 above)
Comment (beckerg) [reply lib/cn/csched_sp3_work.c:427]: done, turned out to be a nightmare...

**Comment by tristan957** [comment]
Regarding commit message, make the summary something like "Avoid k-compactions when they would just be followed immediately by a scatter kv-compaction". Imperative mood!

File: lib/cn/csched_sp3.c (hunk 4 above)
Comment (tristan957) [reply lib/cn/csched_sp3.c:1038]: Yea #defines would be great here with some comments!

File: lib/cn/csched_sp3_work.h [hunk 8]
```diff
@@ -15,6 +15,7 @@
 
 /* clang-format off */
```
Comment (tristan957) [inline lib/cn/csched_sp3_work.h:16]: Why is this here?

Comment (beckerg) [reply lib/cn/csched_sp3_work.h:16]: I moved it up a few lines, it should be format-on...

Comment (beckerg) [reply lib/cn/csched_sp3_work.h:16]: Sorry, it's there because stupid clang format, which is not ready for prime time, kept effing up the spacing in the #defines making them difficult to read.  So same reason you see that anywhere in our code...

Comment (tristan957) [reply lib/cn/csched_sp3_work.h:16]: Right, I am thinking just make it as small a clang-format ignore portion as possible. I didn't see why it is turned off prior to the the struct definitions.

File: lib/cn/kvset.h [hunk 9]
```diff
@@ -355,7 +355,38 @@ kvset_get_compc(struct kvset *km);
...
+ * workaround to avoid calling kvset_get_vgroups().
+ */
+static HSE_ALWAYS_INLINE
+uint
+kvset_get_scatter(const struct kvset *kvset, uint scatter)
+{
+#if !HSE_MOCKING
+    const uint vgroups = kvset_get_vgroups(kvset);
+
+    if (scatter + vgroups > 1)
```
[the remark this answers is not in the document]
Comment (beckerg) [reply lib/cn/kvset.h:381]: The idea is to not count the oldest contiguous kvsets that are fully compacted.  Hence, this function must only be called while scanning a node from oldest to youngest kvset, supplying the previous sum of scatter scores.
So, starting from the oldest kvset, we return 0 until we find a kvset with (vgroups > 1).  At least that's the idea, I rewrote it six times from it's original to make tristan and the unit test framework happy, which then left me very unhappy :-)

File: lib/cn/csched_sp3_work.h (hunk 8 above)
Comment (beckerg) [reply lib/cn/csched_sp3_work.h:16]: Hmm, I think it covers the structs because clang-format wrecks the spacing on the comments?

File: lib/kvdb/kvdb_rparams.c [hunk 10]
```diff
@@ -806,12 +806,12 @@ static const struct param_spec pspecs[] = {
...
         },
         .ps_bounds = {
             .as_uscalar = {
                 .ps_min = 1,
-                .ps_max = 128,
+                .ps_max = 1024,
```
[the remark this answers is not in the document]
Comment (beckerg) [reply lib/kvdb/kvdb_rparams.c:814]: Sigh, thanks!

**Comment by beckerg** [comment]
All checks have passed!

**Comment by tristan957** [comment]
All I will say is when this commit history gets squashed, it is not going to be pretty because the summary will be:

`greg/NFSE-5232-defer-kvcompact (313)`

Would recommend cleaning this up prior to merge.

File: lib/cn/kvset.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (nabeelmmd) [reply lib/cn/kvset.h:381]: Cool, thanks! Was just thinking if we need to differentiate a node with 4 kvsets each having 1 vgroup versus a node with 1 kvset having one vgroup.

From the point of view of a cursor, the former is less preferable as it has to walk 4 different vblock streams, so in the previous generation of scatter remediation we used to generate a node score of 4 in the former case and 1 in the latter case. Maybe these don't matter anymore..

**Review by nabeelmmd — APPROVED** [review]

**Comment by beckerg** [comment]
Not obvious to me how to fix the summary, and not sure how it got that way in the first place..

File: lib/cn/csched_sp3_work.c [hunk 11]
```diff
@@ -429,8 +419,25 @@ sp3_work_leaf_len(
...
-        if (runlen >= runlen_min)
-            return runlen;
+        if (runlen >= runlen_min) {
+
+            /* If scatter is low then simply kcompact the full run of kvsets.
+             */
+            if (scatter < thresh->lscatter_hwm)
+                return runlen;
+
+            /* If the run is longer than the minimum then kcompact just the
```
Comment (nabeelmmd) [inline lib/cn/csched_sp3_work.c:429]: Just clarifying - this means that there could exist kvsets whose vgroup count exceeds the configured vgroups max for a kvdb, right?

**Comment by beckerg** [comment]
Did we recently enable squash to use PR title?  I've been setting the title to the Jira ticket name forever, so that's either been bad or something has changed where it stopped using the commit msg for the summary...

Not clear to me how to "set the description of the PR"...

**Comment by tristan957** [comment]
It has always been like that.

See: https://github.com/hse-project/hse/commit/f2a37908e83f23fc5c55ab321b3b411b1e1167f7

Use the three dot menu at the top of the PR, and hit "Edit"

**Comment by beckerg** [comment]
Looks like 'gh pr -R hse-project/hse edit' does the trick...

File: lib/cn/cn_tree.c [hunk 12]
```diff
@@ -2211,6 +2211,18 @@ cn_tree_node_mclass(struct cn_tree_node *tn, enum hse_mclass_policy_dtype dtype)
...
+cn_tree_node_scatter(const struct cn_tree_node *tn)
+{
+    struct kvset_list_entry *le;
+    uint scatter = 0;
+
+    list_for_each_entry(le, &tn->tn_kvset_list, le_link)
+        scatter += kvset_get_scatter(le->le_kvset, scatter);
+
+    return scatter;
+}
```
Comment (alexttx) [inline lib/cn/cn_tree.c:2224]: A node with 8 kvsets, each with 1 vgroup as scatter == 0.
If the oldest ksvet has 2 vgroups, and all others have 1, then the node scatter is 2.
If the newest kvset has 2 vgroups, and all others have 1, the node scatter is 10.
That doesn't seem right.

Comment (alexttx) [reply lib/cn/cn_tree.c:2224]: Why not simply define node scatter as sum of vgroups - number of kvsets ?  Then the above examples would have scatter of 0, 1 and 1.

Comment (beckerg) [reply lib/cn/cn_tree.c:2224]: I think maybe you're assuming kvset_get_scatter() simply returns the vgroup count, right?  It's not that simplistic...

If oldest kvset has 2 vgroups and all others have 1 then the scatter is 9.
If the newest kvset has 2 vgroups and all other have 1 then scatter is 2.

Searching from the oldest to youngest kvsets, we skip fully compacted kvsets (i.e., vgroups === 1) until we find a kvset with (vgroups > 1), then we sum up all the kvsets of the remaining younger kvsets to get the scatter score.
Prior to patch I didn't sum up the younger kvsets if they were fully compacted.  But now I do so that when the leaf-len rule and scatter rule both see that the scatter score reaches the HWM then the leaf-len rule will defer to the scatter rule until the leaf-len gets excessively long then it will do some smaller fixed k-compacts while waiting for a scatter job to come along.

The scatter rule does try to avoid including the youngest kvsets with no scatter, but it will take all such kvsets above the lrunlen_min threshold (4 kvsets).  We could include them as well, but seems preferable not to turn this job into a full gc operation...

Comment (alexttx) [reply lib/cn/cn_tree.c:2224]: Right.  I just had it backwards and was off by one.  My point is that it seems odd that the order of kvsets would affect the nodes scatter metric.

Comment (beckerg) [reply lib/cn/cn_tree.c:2224]: Now that you made me think about it, there's another k-compaction optimization we can make.  We've long had this problem (looking at compc) where something like 0 0 0 1 1 1 2 2 2 will require three k-compacts in a row to reduce once we add one more new kvset to the left of the list.

But I just realized I can do that in one k-compact with this one simple trick!  (i.e., only reset the mark if the current run length is less than (runlen_min - 1).  Woot!

Comment (alexttx) [reply lib/cn/cn_tree.c:2224]: With this one simple trick!

File: lib/cn/csched_sp3_work.c [hunk 13]
```diff
@@ -326,65 +326,58 @@ sp3_work_leaf_garbage(
...
+            kvset_stats(le->le_kvset, &stats);
 
-            assert(le != *mark);
-            --kvsets;
+            if (stats.kst_kwlen + stats.kst_vwlen < (128ul << 20)) {
+                ev_debug(1);
+                *mark = le;
+                ++kvsets_max;
+                ++kvsets;
+            }
```
Comment (beckerg) [inline lib/cn/csched_sp3_work.c:376]: I'll remove that ev_debug...

**Review by alexttx — APPROVED** [review]
